src/enum.ts

```diff
@@ -1,265 +1,198 @@
-};
-type TrueOrObj<T> = {
-	[K in keyof T]: keyof T[K] extends never ? true : Identity<T[K]>;
-};
 
-/**
-Infers an object mapping variant names to their unit or struct variant values.
-
-```ts
-type Foo = Enum<{ A: true; B: { b: string }; C: { c: number } }>;
-
-Enum.Unwrap<Foo>
--> | { A: true; B: { b: string }; C: { c: number } }
-```
- */
-type EnumUnwrap<TEnum> = Identity<
+export const Enum = <
+	TEnum extends Enum.Any<TDiscriminant>,
+	TMapper extends Partial<
+		Identity<
+			Intersect<
+				TEnum extends unknown
+					? {
+							[Key in TEnum[TDiscriminant]]: (
+								...args: any[]
+							) => Omit<TEnum, TDiscriminant>;
+						}
+					: never
+			>
+		>
+	>,
+	TDiscriminant extends keyof TEnum & string = keyof TEnum &
+		Enum.Discriminant.Default,
+>(
+	_value: TEnum,
+	mapper?: TMapper,
+	discriminant?: TDiscriminant,
+): Identity<
 	Intersect<
-		TEnum extends { is: string }
+		TEnum extends unknown
 			? {
-					[key in TEnum["is"]]: Exclude<keyof TEnum, "is"> extends never
-						? true
-						: Identity<Omit<TEnum, "is">>;
-			  }
+					[Key in TEnum[TDiscriminant]]: Key extends keyof TMapper
+						? TMapper[Key] extends (...args: any[]) => any
+							? (...args: Parameters<TMapper[Key]>) => TEnum
+							: Fn<TEnum, TDiscriminant>
+						: Fn<TEnum, TDiscriminant>;
+				}
 			: never
 	>
->;
-
-/**
-Infers all possible variants' keys of the given Enum.
-
-```ts
-type Foo = Enum<{ A: true; B: { b: string }; C: { c: number } }>;
-
-Enum.Keys<Foo>
--> "A" | "B" | "C"
-```
- */
-type EnumKeys<TEnum> = TEnum extends { is: string } ? TEnum["is"] : never;
-
-/**
-Infers all possible variants' values of the given Enum.
-
-```ts
-type Foo = Enum<{ A: true; B: { b: string }; C: { c: number } }>;
-
-Enum.Values<Foo>
--> | { b: string }
-   | { c: number }
-```
- */
-type EnumValues<TEnum> = TEnum extends { is: string }
-	? Omit<TEnum, "is"> extends infer Values
-		? keyof Values extends never
-			? never
-			: Values
-		: never
-	: never;
-
-/**
-Infers only _common_ variants' properties' names of the given Enum. If `All` is
-`true`, then _all_ variants' properties' names are inferred.
-
-```ts
-type Foo = Enum<{ A: true; B: { x: string }; C: { x: string; y: number } }>;
-
-Enum.Props<Foo>
--> "x"
+> => {
+	return new Proxy({} as any, {
+		get: (_, key: string) => {
+			type LooseMapper = Partial<Record<string, (...args: any[]) => any>>;
+			const dataFn = (mapper as unknown as LooseMapper | undefined)?.[key];
+			return (...args: any[]) => {
+				const data = dataFn ? dataFn(...args) : args[0];
+				return { [discriminant ?? "_type"]: key, ...data };
+			};
+		},
+	});
+};
 
-Enum.Props<Foo, true>
--> "x" | "y"
-```
- */
-type EnumProps<TEnum, TAll extends boolean = false> = [TEnum] extends [never]
-	? never
-	: TAll extends true
-	? TEnum extends unknown
-		? keyof Omit<TEnum, "is">
-		: never
-	: [TEnum] extends [{ is: string }]
-	? keyof Intersect<Omit<TEnum, "is">>
+type Fn<
```

Comment:
I need to rename this because I'm not sure what it's for.

Reply:
Fixed with: https://github.com/peterboyer/unenum/pull/1/commits/892ae52453800ea3aa05ba74aeed72f1aacb2b82.